rmw_fastrtps_shared_cpp/src/rmw_security_logging.cpp

```diff
@@ -0,0 +1,196 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tinyxml2.h>
+
+#include <utility>
+#include <string>
+
+#include "rmw/error_handling.h"
+#include "rmw/qos_profiles.h"
+#include "rmw/types.h"
+
+#include "rmw_fastrtps_shared_cpp/rmw_security_logging.hpp"
+
+namespace
+{
+const char logging_plugin_property_name[] = "dds.sec.log.plugin";
+const char log_file_property_name[] = "dds.sec.log.builtin.DDS_LogTopic.log_file";
+const char verbosity_property_name[] = "dds.sec.log.builtin.DDS_LogTopic.event_log_level";
+const char distribute_enable_property_name[] =
+  "dds.sec.log.builtin.DDS_LogTopic.distribute";
+const char distribute_depth_property_name[] =
+  "com.rti.serv.secure.logging.distribute.writer_history_depth";
+
+const struct
+{
+  const std::string name;
+  rmw_qos_profile_t profile;
+} supported_profiles[] =
+{
+  {"SENSOR_DATA", rmw_qos_profile_sensor_data},
+  {"PARAMETERS", rmw_qos_profile_parameters},
+  {"DEFAULT", rmw_qos_profile_default},
+  {"SERVICES_DEFAULT", rmw_qos_profile_services_default},
+  {"PARAMETER_EVENTS", rmw_qos_profile_parameter_events},
+  {"SYSTEM_DEFAULT", rmw_qos_profile_system_default},
+};
+
+bool string_to_rmw_qos_profile(const std::string & str, rmw_qos_profile_t & profile)
+{
+  for (const auto & item : supported_profiles) {
+    if (item.name == str) {
+      profile = item.profile;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void add_property(
+  eprosima::fastrtps::rtps::PropertySeq & properties,
+  eprosima::fastrtps::rtps::Property && property)
+{
+  // Add property to vector. If property already exists, overwrite it.
+  std::string property_name = property.name();
+  for (auto & existing_property : properties) {
+    if (existing_property.name() == property_name) {
+      existing_property = property;
+      return;
+    }
+  }
+
+  properties.push_back(property);
+}
+
+bool add_property_from_xml_element(
+  eprosima::fastrtps::rtps::PropertySeq & properties, const std::string & property_name,
+  const tinyxml2::XMLElement & element, const char * const tag_name)
+{
+  auto tag = element.FirstChildElement(tag_name);
+  if (tag != nullptr) {
+    const char * text = tag->GetText();
+    if (text == nullptr) {
+      RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "failed to set security logging %s: improper format",
+        tag_name);
+      return false;
+    }
+
+    add_property(properties, eprosima::fastrtps::rtps::Property(property_name, text));
+  }
+
+  return true;
+}
+
+void add_properties_from_qos_profile(
+  eprosima::fastrtps::rtps::PropertySeq & properties, const rmw_qos_profile_t & profile)
+{
+  // TODO(kyrofa): Talk to artivis about what QoS settings will be exposed.
```

Comment:
@artivis have you done enough investigation into the low-level API to know what knobs we can turn, yet? If so we can probably hammer out the design right here. Otherwise this can wait.

Reply:
Not yet, I'll get onto this first thing next week.

Reply:
We can wait for a follow-up PR. This one was getting big anyway.